feat: add `Go to message` action in various places

Nerixyz opened the pull request:
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description
When searching a message, one might want to see it in context to other messages in chat. That's why this PR adds a `Go to message` action to the search context menu.

## Places

- [x] Search window
  * [x] Regular
  * [x] Multiple channels
  * Should the search window close after clicking the action?
- [x] `/mentions`
- [x] User-Card
  * [x] From mentions
  * [x] From reply-thread
  * [x] From search
  * [x] From search in mentions
  * [x] From multisearch
- [x] Reply-Threads
  * [x] From mentions
  * [x] From user-card
  * [x] From search
  * [x] From search in mentions
  * [x] From multisearch
  * [x] From user-card in search in mentions (?)
- [x] Deleted Messages (link to original message)
  * [x] From search
- ~~If a channel has filters enabled, should there be a `Go to message` as well? For example: one view is the regular chat, one only contains the mod messages.~~

Felanbird commented:
Quick immediate thought(s) based on title alone:
- How will this handle multiple splits of the same channel? (global search)

  - Would the fixing of #3861 allow for the approach where the currently focused tab is prioritized, but not cause issues when the duplicate tab is freshly opened. 
 `e.g`: Using /mentions context (maybe noticing it after 5min), opening a new tab with a duplicate split of said channel. attempting to use that new split to search for the message and ending up with a blank search.

dnsge commented:
It would be great to add this functionality to user cards and reply threads, too. 

One small problem that I noticed: if you open a split with not enough messages to scroll, selecting "Go to message" will force scroll the channel view to scroll. Mousing over/doing something to update the channel view will then make it realize that it shouldn't be scrolled, making it jump back.

Nerixyz commented:
> * How will this handle multiple splits of the same channel? (global search)

All channels that contain this message will scroll to the message. `ChannelView::scrollToMessage` should probably include a check that early-outs if the message cannot possibly be in the channel to avoid searching the entire history.

> * Would the fixing of #3861 allow for the approach where the currently focused tab is prioritized, but not cause issues when the duplicate tab is freshly opened?

There could be a check to only go to a message if the tab is focused if I understand the question correctly.

> if you open a split with not enough messages to scroll, selecting "Go to message" will force scroll the channel view to scroll.

Fixed in 4a9ecb4a85a84fdf83f80d07ad84b2ed994de449.

> It would be great to add this functionality to user cards and reply threads, too.

I agree; added _reply-threads_ to my original comment. Though I don't know if that should happen in this PR.

Felanbird commented on CHANGELOG.md:
```diff
@@ -22,7 +22,8 @@
 - Minor: Added ability to execute commands on chat messages using the message context menu. (#3738, #3765)
-- Minor: Added `/copy` command. Usage: `/copy <text>`. Copies provided text to clipboard - can be useful with custom commands. (#3763)
+- Minor: Added `/copy` command. Usage: `/copy <text>`. Copies provided text to clipboard - can be useful with custom
+  commands. (#3763)
```
🤔

Nerixyz replied:
Oops I accidentally hit enter ![](https://static-cdn.jtvnw.net/emoticons/v2/934643/default/dark/1.0).

pajlada commented on src/widgets/helper/SearchPopup.cpp:
```diff
@@ -106,6 +106,14 @@ void SearchPopup::addChannel(ChannelView &channel)
+{
+    for (const auto &channel : this->searchChannels_)
+    {
+        channel.get().scrollToMessage(message);
```
Early out

Nerixyz replied:
Done in 4ed3937dc6e3351be95858a887720418147b5a32.

pajlada commented on src/widgets/helper/ChannelView.hpp:
```diff
@@ -208,6 +227,12 @@ class ChannelView final : public BaseWidget
+     * @param layout Must be from this channel.
+     * @param messageIdx Must be an index into this channel.
+     */
+    void scrollToMessage(MessageLayout *layout, size_t messageIdx);
```
Rename to something more internally appropriate, maybe something like `scrollToMessageInLayout`

Nerixyz replied:
Done in c8325c32479116991f940b240be69706344abac0.

pajlada commented on CHANGELOG.md:
```diff
@@ -32,6 +32,8 @@
 - Minor: Add AutoMod message flag filter. (#3938)
 - Minor: Added whitespace trim to username field in nicknames (#3946)
+- Minor: Added `Go to message` context menu action to search popup, mentions, user-card and reply-threads (#3953)
+- Minor: Added link to deleted message (#3953)
```
Explain the "deleted message" changelog entry a bit more verbosely

pajlada reviewed (changes requested):
Currently doesn't switch tabs if the message/channel is found in another tab

Other than that, animation and functionality seems great!

Nerixyz commented on CHANGELOG.md:
```diff
@@ -32,6 +32,8 @@
 - Minor: Add AutoMod message flag filter. (#3938)
 - Minor: Added whitespace trim to username field in nicknames (#3946)
+- Minor: Added `Go to message` context menu action to search popup, mentions, user-card and reply-threads (#3953)
+- Minor: Added link to deleted message (#3953)
```
Done in 20fd8ba2d08547c72b3087ad05df13ca4af4d9be.

Nerixyz commented on src/messages/Link.hpp:
```diff
@@ -25,6 +25,7 @@ struct Link {
         CopyToClipboard,
         ReplyToMessage,
         ViewThread,
+        MessageId,
```
Done in d78503937fa0330c23f6326672ce2c64d008256c.

Nerixyz commented on src/singletons/WindowManager.cpp:
```diff
@@ -320,6 +320,11 @@ void WindowManager::select(SplitContainer *container)
     this->selectSplitContainer.invoke(container);
 }
 
+void WindowManager::selectAndScrollToMessage(const MessagePtr &message)
```
Done in 51d231f5099ce8293f0fb169fad3966fcf0cf43b.

pajlada commented on src/widgets/Notebook.cpp:
```diff
@@ -1006,6 +1007,29 @@ SplitNotebook::SplitNotebook(Window *parent)
+                if (auto sc = dynamic_cast<SplitContainer *>(item.page))
+                {
+                    auto &&splits = sc->getSplits();
+                    auto split = std::find_if(
+                        splits.begin(), splits.end(), [message](Split *split) {
+                            return split->getChannel()->getType() !=
+                                       Channel::Type::TwitchMentions &&
+                                   split->getChannelView().scrollToMessage(
+                                       message);
```
Convert this to a for loop instead, remember you'll need to return to use `return` since `break` will only break the inner loop

Nerixyz replied:
Done in 4798ed67d9a77be42a2afd01c74bc3a58cacee37.

pajlada commented on src/widgets/helper/ChannelView.cpp:
```diff
@@ -339,9 +359,26 @@ void ChannelView::themeChangedEvent()
+void ChannelView::setupHighlightAnimationColors()
+{
+    constexpr int maxOpacity = 110;
+    if (this->theme->isLightTheme())
+    {
+        this->highlightAnimation_.setStartValue(QColor(20, 20, 20, maxOpacity));
+        this->highlightAnimation_.setEndValue(QColor(20, 20, 20, 0));
+    }
+    else
+    {
+        this->highlightAnimation_.setStartValue(
+            QColor(230, 230, 230, maxOpacity));
+        this->highlightAnimation_.setEndValue(QColor(230, 230, 230, 0));
+    }
```
Instead of checking for light theme, define two colors (`highlightAnimationStart` and `highlightAnimationEnd`) in the theme class that does the light theme checking

That way this class would not need a branch, and just reference the two new colors

Nerixyz replied:
Done in aed9164c0b420ce7c937f21507db91e3bf7cef5f.

pajlada commented on src/widgets/helper/ChannelView.hpp:
```diff
@@ -83,6 +84,17 @@ class ChannelView final : public BaseWidget
     const boost::optional<MessageElementFlags> &getOverrideFlags() const;
     void updateLastReadMessage();
 
+    /**
+     * Attempts to scroll to a message in this channel.
+     * @return <code>true</code> iff the message was found and highlighted.
+     */
+    bool scrollToMessage(const MessagePtr &message);
+    /**
+     * Attempts to scroll to a message id in this channel.
+     * @return <code>true</code> iff the message was found and highlighted.
+     */
+    bool scrollToMessageId(const QString &id);
```
Just use `if` instead of `iff` here

Nerixyz replied:
Done in df19120fe14c32ff7f125e48e8ceda399edd01d5.

pajlada commented on src/widgets/Notebook.cpp:
```diff
@@ -1015,17 +1015,14 @@ SplitNotebook::SplitNotebook(Window *parent)
+                        if (split->getChannel()->getType() !=
+                            Channel::Type::TwitchMentions)
+                        {
+                            split->getChannelView().scrollToMessage(message);
+                            return;
```
Should only return if `scrollToMessage` returns true

Nerixyz replied:
Done in 8ed335105a8d3c07791d3251281bb546b78dd2bb.
